### PR TITLE
[AUTHZ] Check previleges for Iceberg's CALL to RewriteDataFilesProcedure procedure

### PR DESCRIPTION
**Pull request by yabola**
### _Why are the changes needed?_

Before this PR, Kyuubi skips authentication for the call command in iceberg. After this PR, it needs `alter` permission to use this command.
This PR supports checking iceberg call rewrite_data_files command:
- CALL ice.system.rewrite_data_files(table => 'ice.lcl.test_ice', options => map('min-input-files','2'))
- some other call statements in iceberg ( https://iceberg.apache.org/docs/latest/spark-procedures ) may be supported but needs to be further determined. The next phase is planned to support all commands as documented.

When the `rewrite_data_files` command is called, there are two situations:
1. When the triggering conditions are not be met(such as input-files < min-input-files) , there is only one logical plan.  We  need to check one logical plan permission:
```
        * == Physical Plan 1 ==
       * Call (1)
```
2. When the triggering  conditions are met, two logical plan will be triggered. We should check two logical plans permission.
```
       * == Physical Plan 1 ==
       * Call (1)
       *
       * == Physical Plan 2 ==
       * AppendData (3)
       * +- * ColumnarToRow (2)
       * +- BatchScan local.iceberg_ns.call_command_table (1)
```
       
The second plan (`AppendData`) is a bit special,  causing us to have to modify the logic of extracting table names in `DataSourceV2RelationTableExtractor`. The `DataSourceV2Relation#identifier` in `AppendData` will be rewrite task id, so it is a more appropriate way to parse the table name from `DataSourceV2Relation#table`
image

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/contributing/code/testing.html#running-tests) locally before make a pull request

### _Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by pan3793** [comment]
@yabola thanks for your contribution. You may want to send a minor PR(e.g. bump iceberg from 1.3.0 to 1.3.1) first to unblock the CI (auto-run without approval)

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/tableExtractors.scala
Comment (yabola) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/tableExtractors.scala:148]: The `identifier` in `DataSourceV2Relation` is not reliable enough. In fact, datasourceV2 will parse out the table based on identifier, so table is more appropriate here.
For example, in iceberg `rewrite_data_files` command, `identifier` is rewrite task id, iceberg catalog will load table by it ( can see in `SparkCatalog#load` the `isPathIdentifier` function)

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/tableExtractors.scala
Comment (yabola) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/tableExtractors.scala:146]: iceberg will rearrange the parameters according to the parameter order defined in the procedure, where the table parameters are currently always the first.
For example `RewriteDataFilesProcedure#PARAMETERS` and in `RewriteDataFilesProcedure#call` take the first parameter.

**Comment by bowenliang123** [comment]
Hi @yabola, thanks for the contribution to the Authz plugin. And pleased to see introducing support for CALL statements for Iceberg which are critical to lakehouse lifecycle.

Some advices here,
1. Move the command spec generation to `authz.gen` package's `IcebergCommands`, which then is used in `JsonSpecFileGenerator` for spec json file generation by running `dev/gen/gen_ranger_spec_json.sh` locally . And also running `KYUUBI_UPDATE=0 dev/gen/gen_ranger_spec_json.sh` to check spec file.
2. if possible, try not to introduce a new TableExtractor specifically for Iceberg. Iceberg commands are generally supported with existed extractors.

**Comment by yabola** [comment]
@bowenliang123 Thank you for your advice.
1. I add new method in `TableCommands` and generate json by `KYUUBI_UPDATE=1 build/mvn clean test -Pgen-policy -pl :kyuubi-spark-authz_2.12 -Dtest=none -DwildcardSuites=org.apache.kyuubi.plugin.spark.authz.gen.JsonSpecFileGenerator` command. Please see if it is suitable.
2. I removed one unnecessary table extractor. But there may a new table extractor may need to be added because the iceberg call command is a new logical plan(`Call(procedure: Procedure, args: Seq[Expression])`). Or do you have a better suggestion?

**Comment by bowenliang123** [comment]
1. Please move `IcebergCallCommand ` to `IcebergCommands` class as mentioned above, as `org.apache.spark.sql.catalyst.plans.logical.Call` is from Iceberg but not a general plan from Spark engine. And result spec file is check by CI tests.
2. It's feasible to introduce a new table extractor for the Call plan.
3. Try not to introduce `resolveIcebergTable` in `tableExtractors`, and do the extraction by reusing existed extractors following the main procedures. Reversed coupling is always unwelcomed. Any other lakehouse may and will be supported by design with in separated command extractors and specs, while sharing the same common structure.

**Comment by yabola** [comment]
1. I have moved to `IcebergCallCommand`. Sorry didn't notice...
3. Yes, there should share the same common structure. I think we'd better use the `table` parameter instead of `identifier` here. The reason:
> The identifier in DataSourceV2Relation is not reliable enough. In fact, datasourceV2 will parse out the table based on identifier, so table is more appropriate here.
For example, in iceberg rewrite_data_files command, identifier is rewrite task id, iceberg catalog will load table by it ( can see in SparkCatalog#load the isPathIdentifier function)

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/IcebergCommands.scala
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/IcebergCommands.scala:54]: `UPDATE` is the more proper one for the action type.

Comment (yabola) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/gen/IcebergCommands.scala]: Thanks, done

**Comment by bowenliang123** [comment]
Pleased to see the progress of the added e2e tests which passed CI tests.
Still some concerns in the following changes, that I may have further investigation or proof.
1. Again, change the source field inside the `DataSourceV2RelationTableExtractor`, from "identifier" to `Table` "table"
2. the introduced `StringTableExtractor` 
3. forced casting to Literal in `IcebergCallArgsTableExtractor `

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/IcebergCatalogPrivilegesBuilderSuite.scala
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/IcebergCatalogPrivilegesBuilderSuite.scala:145]: The SELECT query access privilege of the source table should be required when calling the table rewriting produce.

**Comment by yabola** [comment]
@bowenliang123 Thank you for the improvement

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/tableExtractors.scala
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/tableExtractors.scala:148]: Oh, got your point. Iceberg procedure rewrites this identifier, and makes it irrelevant to the original table.

Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/tableExtractors.scala:148]: I think we need to `TableTableExtractor` for taking `Table` input, instead of use `StringTableExtractor` directly.

Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/tableExtractors.scala:148]: Please attach the evidence showing the impacts and differences in logical plan details of Iceberg call procedure.

Comment (yabola) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/serde/tableExtractors.scala:148]: When CALL ice.system.rewrite_data_files command does the rewrite operation, two logical plans will be generated, for example in my unit test:

> 
  * This triggers two logical plans( input-files(2) >= min-input-files(2) ):
   *
   * == Physical Plan 1 ==
   * (1) Call
   *
   * == Physical Plan 2 ==
   * AppendData (3)
   * +- * ColumnarToRow (2)
   * +- BatchScan local.iceberg_ns.call_command_table (1)

 In the second logical plan, the `DataSourceV2Relation#identifier` in `AppendData` will be rewrite task id. So it is a more appropriate way to parse the table name from `DataSourceV2Relation#table`
 
image

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala:238]: We'd better use RewriteDataFilesProcedure, refering to `org.apache.iceberg.spark.procedures.RewriteDataFilesProcedure` from Iceberg Spark plugin.

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala:256]: Let's skip testing on `CALL system.set_current_snapshot` procedure this time. Move to the follow-ups.

**Comment by yabola** [comment]
@bowenliang123 Thank you for adding new ranger policy for UT and the relevant code improvements. And is UT failure due to instability?

**Comment by bowenliang123** [comment]
> @bowenliang123 Thank you for adding new ranger policy for UT and the relevant code improvements. And is UT failure due to instability?

Hi, updated the table spec json just now. I changed the operation time from QUERY to ALTERTABLE_ADDPARTS, which is not a best one but a closer one.

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala:288]: Could you attach the triggered pysical plan for this situation as well? @yabola

Comment (yabola) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala:288]: I have added the physical plan.

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala:281]: Do you mean 'result to 2 rewritten data files' or 'exited 2 files before/after procedure'?

Comment (yabola) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala:285]: It means number of data files which were re-written by this command.  The explanation of the output of this command can refer to https://iceberg.apache.org/docs/latest/spark-procedures/#rewrite_data_files

**Comment by bowenliang123** [comment]
"Permission denied: user [someone] does not have [alter] privilege on [iceberg_ns/table_select_call_command_table]"
 did not contain 
"does not have [select] privilege on [iceberg_ns/table_select_call_command_table], [update] privilege on [iceberg_ns/table_select_call_command_table]"

PrivilegesBuilder extracted two privilege object for input and output.
But their action type both are `PrivilegeObjectActionType.OTHER` (Possible value of PrivilegeObjectActionType are OTHER, INSERT, INSERT_OVERWRITE, UPDATE, DELETE).
Therefore, RuleAuthorization would then treat them as the same privilege object, and check the `alter` privilege only.

1.  it seems we could not distinguish them, by 1 for `select` on input table and 1 for `alter` on output table. cc @yaooqinn 
2. Or let's check alter privilege only, and skip checking `select` privilege, as the same and only table used in one procedure. @yabola

File: extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/test/scala/org/apache/kyuubi/plugin/spark/authz/ranger/IcebergCatalogRangerSparkExtensionSuite.scala]: OK, thanks. Please mention these details in the description.

**Comment by yabola** [comment]
> ```
> "Permission denied: user [someone] does not have [alter] privilege on [iceberg_ns/table_select_call_command_table]"
>  did not contain 
> "does not have [select] privilege on [iceberg_ns/table_select_call_command_table], [update] privilege on [iceberg_ns/table_select_call_command_table]"
> ```
> 
> PrivilegesBuilder extracted two privilege objects for input and output. But their action type both are `PrivilegeObjectActionType.OTHER` (Possible values of PrivilegeObjectActionType are OTHER, INSERT, INSERT_OVERWRITE, UPDATE, DELETE). Therefore, RuleAuthorization would then treat them as the same privilege object, and check the `alter` privilege only.
> 
> 1. it seems we could not distinguish them, by 1 for `select` on input table and 1 for `alter` on output table. cc @yaooqinn
> 2. Or let's check alter privilege only, and skip checking `select` privilege, as the same and only table used in one procedure, and no data changed by calling the procedure. @yabola

I think having a single authority is more direct and easier for people to understand.

**Review by bowenliang123 — APPROVED** [review]
LGTM.
All Authz tests passed on the local machine.

**Comment by bowenliang123** [comment]
Thank you for your contribution and patience. Looking forward to your follow-up PRs in improving Authz support for Iceberg.
Merged to master (1.9.0).

**Comment by elisabetao** [comment]
Following the TableExtractor adjustments it seems the Iceberg metadata tables (snapshots,manifests,history) are being validated for authorization . Is this expected?  Thanks a lot or the help and feedback

**Comment by bowenliang123** [comment]
> Following the TableExtractor adjustments it seems the Iceberg metadata tables (snapshots,manifests,history) are being validated for authorization . Is this expected? Thanks a lot or the help and feedback

#3931 may help, but it has not been accepted.

**Pull request closed without merging**
